guardian/aws-canaries
# Automate canary deployment

## What does this change?

This enables Continuous Deployment for our AWS Canaries. When we push a change to this repo, the canaries will now be automatically deployed to AWS using Riffraff, which means we no longer have the manual process of copy and pasting the code in to the AWS console.

## New process

We use a combination of Github Actions and Riffraff.

On a push to the main branch, the Github action deploy.yaml does the following:
- Checks out code and installs node
- Runs `create-artifacts.sh`, which zips up our lambda functions. These need to be zipped, as AWS expects to find a zip file containing a template when creating a canary.
- Uses Guardian CDK to create CloudFormation templates
- Uploads each zip file and CloudFormation template to Riffraff using [actions-riff-raff](https://github.com/guardian/actions-riff-raff)
- Gives instructions to Riffraff on uploading lambda functions to S3 and how to run the CloudFormation template to create resources.

Continuous deployment is set up in Riffraff for each AWS region. Riffraff will then upload the files to S3 and execute the CloudFormation script to create the necessary resources.

## How to test

Create a branch, make a change to the code and push to origin. Your code will be automatically deployed to a test canary. Check that your code is reflected in AWS Synthetics Canaries. The test canary is called `comm_cmp_canary_code`. If you had pushed to main, the prod canary `comm_cmp_canary_prod` would have instead been updated.

## The following resources are created

- Synthetics canary
- IAM role
- IAM policies
- CloudWatch Alarm (PROD only)
- SNS Topic (PROD only)
- Lambda with layers (not specified in yaml file: deployed automatically by AWS)
- S3 bucket (not specified in yaml file: already exists)

## Next Steps 

#### Post-PR review

- Before merging to main, we need to manually delete the existing PROD lambda resources. This is because merging to main will trigger CD which will run the CF template to create new resources, which will then fail because the existing canary and the new canary have the same name.

#### Post-merge to main

- Monitor the logs after the CD is running to make sure that these are written correctly.

## Future (clean-up and extra tasks)

- Tidy up unused buckets. We currently have canaries and canary logs in buckets that we no longer use. After 30 days, logs will be deleted and these buckets will no longer contain any useful information. Candidates are "aws-commercial-canaries" and all those beginning "cw-syn-results-". Follow steps here: https://github.com/guardian/recommendations/blob/main/AWS.md#s3
- Clean up IAM roles. We currently have >20 roles with "CloudWatchSyntheticsRole" in the name. We only need 8 (2 for each region: 1 Prod, 1 Test).
- Include S3 bucket in the template. Note that we do not want the bucket to be deleted if the CF stack is deleted: it contains the canary logs
- Update synthetics to v3.6 (A one-liner: a good candidate to test that the whole CD process works)
